From d33f778ca4992fc2a38c2f72dd9c16bb1e4512af Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:37:09 -0400
Subject: [ethernet] i40e: remove open-coded skb_cow_head

Message-id: <1409658042-22689-69-git-send-email-sassmann@redhat.com>
Patchwork-id: 90263
O-Subject: [RHEL7.1 PATCH 068/281] i40e: remove open-coded skb_cow_head
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Francois Romieu <romieu@fr.zoreil.com>

Signed-off-by: Francois Romieu <romieu@fr.zoreil.com>
Cc: Jesse Brandeburg <jesse.brandeburg@intel.com>
Tested-by: Kavindya Deegala <kavindya.s.deegala@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit dd225bc675bc16972cc11f73fa0dc3ccb1ed9da1)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_txrx.c b/drivers/net/ethernet/intel/i40e/i40e_txrx.c
index 0f5d96a..1fdc8e9 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_txrx.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_txrx.c
@@ -1713,9 +1713,11 @@ static int i40e_tx_prepare_vlan_flags(struct sk_buff *skb,
     I40E_TX_FLAGS_VLAN_PRIO_SHIFT;
   if (tx_flags & I40E_TX_FLAGS_SW_VLAN) {
    struct vlan_ethhdr *vhdr;
-   if (skb_header_cloned(skb) &&
-       pskb_expand_head(skb, 0, 0, GFP_ATOMIC))
-    return -ENOMEM;
+   int rc;
+
+   rc = skb_cow_head(skb, 0);
+   if (rc < 0)
+    return rc;
    vhdr = (struct vlan_ethhdr *)skb->data;
    vhdr->h_vlan_TCI = htons(tx_flags >>
        I40E_TX_FLAGS_VLAN_SHIFT);
@@ -1743,20 +1745,18 @@ static int i40e_tso(struct i40e_ring *tx_ring, struct sk_buff *skb,
       u64 *cd_type_cmd_tso_mss, u32 *cd_tunneling)
 {
  u32 cd_cmd, cd_tso_len, cd_mss;
+ struct ipv6hdr *ipv6h;
  struct tcphdr *tcph;
  struct iphdr *iph;
  u32 l4len;
  int err;
- struct ipv6hdr *ipv6h;
 
  if (!skb_is_gso(skb))
   return 0;
 
- if (skb_header_cloned(skb)) {
-  err = pskb_expand_head(skb, 0, 0, GFP_ATOMIC);
-  if (err)
-   return err;
- }
+ err = skb_cow_head(skb, 0);
+ if (err < 0)
+  return err;
 
  if (protocol == htons(ETH_P_IP)) {
   iph = skb->encapsulation ? inner_ip_hdr(skb) : ip_hdr(skb);
-- 
1.7.1